### PR TITLE
Attackby destruction

### LINKED ISSUE #17122: No cooldown in attacking drone shell dispenser
Hopefully fixed by #17092.

Currently attacking drone shell dispensers has no cooldown, and can do attacks as fast as you can click.

### PR DESCRIPTION
**Pull request by phil235**
- Hitting all machines now make a sound.
- Many machines can now be broken (but not always destroyed) by melee attack/bullet/mech punches/other mob's attacks.
- You can now hit closet, tables and airlock with a weapon by selecting harm intent. But only the table is actually breakable.
- You can hit a lot more structures as well.
- I moved the code handling structural damage of all previously breakable objects out of attackby(). Attackby() now always calls return ..(). The code handling hitting structures/machines is now attacked_by, (very similar to how it's done for mobs: attackby -> Item.attack() -> attacked_by() ; but the second proc is called attack_obj()). The code handling actually giving damage to these objects is all in the take_damage() proc (instead of being scattered in all the attackbys).
- I made door code more OOP. There was a lot of duplication between airlock, firedoor, poddoor, windowdoor.
- Fixes wooden barricade not dropping wood when destroyed.
- Fixes some bug when a mech with armor booster module (Ranged Weaponry) is hit by a rapid bullet burst (sleep() was delaying the mech registering the hits).
- You can now burst water balloons with a sharp item.
- Fixes slot machine being undeconstructable when it has coins in it.

List of machines and structures breakable by projectiles/mech melee/weapon melee/alien attack/animal attack, explosion:
- PDA painter
- computer (now also with melee weapon)
- DroneDispenser (now also with bullets/animal attacks)
- newscaster (now also with bullets/animal attacks)
- fire alarm
- anti-breach shielding projector
- shield generator
- canister (now also with mob attacks)
- energy net (now also with bullets)
- lights (now also with bullets)
- solar panel (now also with mob attacks)
- solar panel control
- fireaxe cabinet (now also with mob attacks)
- mineral statues (now also with mob attacks)
- table
- rack (now also with melee weapons and bullets)
- broken APC (now also mob attacks and bullets)

fixes #17122
:cl: phil235
rscadd: More structures and machines are now breakable and/or destroyable with melee weapon. Additionally, Some structures and machines that are already breakable/destroyable are now affected by more attack methods (gun projectiles, thrown items, melee weapons, xeno/animal attacks, explosions).
rscadd: Hitting any machine now always makes a sound.
rscadd: You can hit closed closets/crates and doors by using an item on harm intent (but it doesn't actually hurt them).
rscadd: you can now burst filled water balloons with sharp items.
bugfix: Wooden barricade drops wood when destroyed.
bugfix: Slot machines with coins inside are deconstructable now.
/:cl:

### CODE REVIEW

**Comment by AnturK** [comment]
This is scary. But probably good.

File: code/datums/wires/airlock.dm [hunk 1]
```diff
@@ -19,7 +19,7 @@
 	var/obj/machinery/door/airlock/A = holder
 	if(!istype(user, /mob/living/silicon) && A.isElectrified() && A.shock(user, 100))
 		return FALSE
-	if(A.p_open)
+	if(A.panel_open)
```
Comment (phil235) [inline code/datums/wires/airlock.dm]: Now all machines use the same var "panel_open"

**Comment by Dorsisdwarf** [comment]
Does this let you smash open vending machines?

**Comment by phil235** [comment]
@Dorsisdwarf No, I didn't make all machines breakable because: it would've made the PR too big, and there's some balance issue that needs to be addressed for many machines.

**Comment by Jalleo** [comment]
As i said in the commented version we may need to change the health of items to balance it out.
And also the mention about alien_attack proc rng with apcs.

**Comment by phil235** [comment]
@Jalleo Of course, balancing health will be needed. I don't think I've touched APC damage tough, I didn't want to make a sensitive machine easily breakable in this PR.

**Comment by Jalleo** [comment]
Phil it would be better to make it breakable because i believe it already is with rng with quite a few items.

**Comment by phil235** [comment]
@Jalleo Done. It's not fully breakable, but knocking down the cover of a broken APC is now no longer random (and other mob attacks can knock it down as well).

**Comment by AnturK** [comment]
Nothing can go wrong here. :shipit:

**Comment by TechnoAlchemisto** [comment]
You can't place objects on tables to tablecraft

**Comment by phil235** [comment]
@TechnoAlchemisto Are you sure you're not just on harm intent?

**Comment by TechnoAlchemisto** [comment]
I am positive @phil235

**Comment by phil235** [comment]
@TechnoAlchemisto were you a human? What item were you trying to place on the table?

**Comment by TechnoAlchemisto** [comment]
It works now oddly enough, at the time I was on help intent and putting bones onto the survival pod table in mining

**Pull request merged**